pkg/asset/quota/openstack/openstack.go

```diff
@@ -11,44 +14,90 @@ import (
+	}
+	constraints = append(constraints, instanceConstraint(int64(len(controlPlanes))))
+
+	for i := 0; i < len(computes); i++ {
+		constraints = append(constraints, machineSetConstraints(ci, &computes[i])...)
+	}
+	constraints = append(constraints, instanceConstraint(int64(len(computes))))
+
+	return aggregate(constraints)
+}
+
+func getOpenstackProviderSpec(spec *machineapi.ProviderSpec) *openstackprovider.OpenstackProviderSpec {
+	// An invalid ProviderSpec would presumably be a programming error here.
+	if spec.Value == nil {
+		logrus.Warnf("Empty ProviderSpec")
+		return nil
 	}
-	computeReplicas := make([]int64, len(computes))
-	computeConfigs := make([]*openstackprovider.OpenstackProviderSpec, len(computes))
-	for i, m := range computes {
-		computeReplicas[i] = int64(*m.Spec.Replicas)
-		computeConfigs[i] = m.Spec.Template.Spec.ProviderSpec.Value.Object.(*openstackprovider.OpenstackProviderSpec)
+
+	osps, ok := spec.Value.Object.(*openstackprovider.OpenstackProviderSpec)
+	if !ok {
+		logrus.Warnf("ProviderSpec of unexpected type %T", spec.Value.Object)
+		return nil
 	}
 
-	var ret []quota.Constraint
-	ret = controlPlane(ci, ctrplConfigs)
-	ret = append(ret, compute(ci, computeReplicas, computeConfigs)...)
-	return aggregate(ret)
+	return osps
 }
 
-func controlPlane(ci *validation.CloudInfo, machines []*openstackprovider.OpenstackProviderSpec) []quota.Constraint {
-	var ret []quota.Constraint
-	for _, m := range machines {
-		flavorInfo := ci.Flavors[m.Flavor]
-		ret = append(ret, machineFlavorCoresToQuota(flavorInfo, ci), machineFlavorRAMToQuota(flavorInfo, ci))
+func getFlavor(ci *validation.CloudInfo, flavorName string) *flavors.Flavor {
+	if flavorInfo, ok := ci.Flavors[flavorName]; ok {
+		return flavorInfo.Flavor
 	}
-	ret = append(ret, instanceConstraint(int64(len(machines))))
-	return ret
+	return nil
 }
 
-func compute(ci *validation.CloudInfo, replicas []int64, machines []*openstackprovider.OpenstackProviderSpec) []quota.Constraint {
-	var ret []quota.Constraint
-	for idx, m := range machines {
-		flavorInfo := ci.Flavors[m.Flavor]
-		coresConstraint := machineFlavorCoresToQuota(flavorInfo, ci)
-		coresConstraint.Count = coresConstraint.Count * replicas[idx]
-		ramConstraint := machineFlavorRAMToQuota(flavorInfo, ci)
-		ramConstraint.Count = ramConstraint.Count * replicas[idx]
-		ret = append(ret, instanceConstraint(int64(replicas[idx])), coresConstraint, ramConstraint)
+func machineConstraints(ci *validation.CloudInfo, machine *machineapi.Machine) []quota.Constraint {
+	osps := getOpenstackProviderSpec(&machine.Spec.ProviderSpec)
+	if osps == nil {
+		logrus.Warnf("Skipping quota validation for Machine %s: Invalid ProviderSpec", machine.Name)
+		return nil
+	}
+
+	flavor := getFlavor(ci, osps.Flavor)
+
+	// This will result in a separate validation failure
+	if flavor == nil {
+		logrus.Warnf("Skipping quota validation for Machine %s: Flavor '%s' is not valid",
+			machine.Name, osps.Flavor)
+		return nil
```

Comment:
this block goes into the `!ok` arm of the standard Go map lookup:

```Go
var flavor *flavors.Flavor
if flavor, ok := ci.Flavors[flavorName]; !ok {
	logrus.Warnf("Skipping quota validation for Machine %s: Flavor '%s' is not valid",
	machine.Name, osps.Flavor)
	return nil
}
return []quota.Constraint{machineFlavorCoresToQuota(flavor), machineFlavorRAMToQuota(flavor)}
```

Reply:
If that's more idiomatic, then sure. How about using an else rather than defining the variable outside of the scope of the if statement? They're both a bit ugly, tbh.

Reply:
Defining the variable in the outer scope lets you apply early returns, which is another fashionable way to show off your Go orthodoxy